o1vm/src/mips/witness.rs

```diff
@@ -667,57 +660,85 @@ impl<Fp: Field, PreImageOracle: PreImageOracleT> InterpreterEnv for Env<Fp, PreI
                 self.write_column(
-                    Column::ScratchState(MIPS_PREIMAGE_BYTES_OFFSET + i as usize),
+                    Column::ScratchState(MIPS_PREIMAGE_BYTES_OFF + byte_i),
                     preimage_byte as u64,
                 );
+
+                // Update the chunk of at most 4 bytes read from the preimage
+                chunk = chunk << 8 | preimage_byte as u64;
+
+                // At most, it will be actual_read_len when the length is not
+                // read in this call
+                preimage_read_len += 1;
+
+                // TODO: Proabably, the scratch state of MIPS_PREIMAGE_BYTES_OFF
+                // is redundant with lines below
                 unsafe {
                     self.push_memory(&(*addr + i), preimage_byte as u64);
                     self.push_memory_access(&(*addr + i), self.next_instruction_counter());
                 }
             }
         }
+        // Update the chunk of at most 4 bytes read from the preimage
+        // FIXME: this is not linked to the registers content in any way.
+        //        Is there anywhere else where the bytes are stored in the
+        //        scratch state?
+        self.write_column(Column::ScratchState(MIPS_PREIMAGE_CHUNK_OFF), chunk);
+
+        // Update the number of bytes read from the oracle in this step (can
+        // include bytelength and preimage bytes)
         self.write_column(pos, actual_read_len);
 
+        // Number of preimage bytes processed in this instruction
+        self.write_column(
+            Column::ScratchState(MIPS_NUM_BYTES_READ_OFF),
+            preimage_read_len,
+        );
+
         // Update the flags to count how many bytes are contained at least
-        for i in 0..MIPS_CHUNK_BYTES_LENGTH {
+        for i in 0..MIPS_CHUNK_BYTES_LEN {
             if preimage_read_len > i as u64 {
-                self.write_column(Column::ScratchState(MIPS_HAS_N_BYTES_OFFSET + i), 1);
+                // This amount is only nonzero when it has read some preimage
+                // bytes.
+                self.write_column(Column::ScratchState(MIPS_HAS_N_BYTES_OFF + i), 1);
             }
         }
 
         // Update the total number of preimage bytes read so far
         self.preimage_bytes_read += preimage_read_len;
         self.write_column(
-            Column::ScratchState(MIPS_BYTES_READ_OFFSET),
+            Column::ScratchState(MIPS_BYTE_COUNTER_OFF),
             self.preimage_bytes_read,
         );
 
-        // Update how many bytes are left to be read
-        self.write_column(
-            Column::ScratchState(MIPS_PREIMAGE_LEFT_OFFSET),
-            (preimage_len as u64) - self.preimage_bytes_read,
-        );
-
         // If we've read the entire preimage, trigger Keccak workflow
         if self.preimage_bytes_read == preimage_len as u64 {
+            self.write_column(Column::ScratchState(MIPS_END_OF_PREIMAGE_OFF), 1);
```

Comment:
Even though the default value is 0 for the scratch state, I would also write 0 then in the other case.